### DISCUSSION TITLE
Fits/InputDecoder Bulk Reads

### DISCUSSION THREAD

**Discussion by keastrid** [nom-tam-fits/nom-tam-fits#413] [Q&A] [answered]
I might be misunderstanding something, but it seems like the bulk read methods in `FitsDecoder` such as `read(float[],...)` go element by element when reading. 

Is there a reason `InputBuffer` doesn't create type specific views such as `FloatBuffer` and use their bulk read methods? It seems to do so for byte arrays.

**Comment by attipaci** [comment]
Not quite...

The short answer is that the FITS reading is probably about as fast as it will ever get. It is primarily limited by the underlying IO performance, even for SSDs. (I benchmarked 800 MB/s read for FITS `double[]` vs 4000 MB/s for  direct`byte[]` on my Gen 3 M.2 laptop SSD and a 6-yo CPU -- not bad given the extra processing needed to byte-swap the `double` values). Now the long answer...

The way nom-tam-fits is implemented, there is always a bulk read of bytes from the underlying input to a `ByteBuffer` (96 kB) when you use either `FitsFile` (or some other extension of `BufferedFileIO`) or `FitsInputStream` (based on `BufferedInputStream`). That ensures that the actual IO part (the slowest part!) of reading -- from storage to RAM -- happens as efficiently as possible, especially for ordered (sequential) access of the FITS data.

But then, to access data from the buffer there is a filter on top, because Java types don't map exactly to FITS types (most notably `boolean` in Java is single bit, whereas in FITS it's a byte with 'T' or 'F'; and characters which are 2-byte unicode in Java but single byte ASCII in FITS). Also because while `ByteBuffer` has methods for accessing elements of all types, it does not have bulk read into arrays of all types (except `byte[]` arrays), so one must iterate over elements for the various types no matter. (See also my note at the bottom why views don't help in this regard -- which is probably the reason also why `ByteBuffer` does not offer bulk read to primitive arrays even though it could if there was any performance advantage to it).

While the translation from Java to FITS types, and the handling of arrays of all types, does happen element by element (which is what `FitsDecoder` does) it is nevertheless very fast since `FitsDecoder` is a accessing buffered data in memory (and usually from the CPU cache[!] in case of sequential reads), not from the input directly. `FitsDecoder` also implements a second layer of buffering (2880 bytes -- i.e. a FITS block) that is decoupled from the input's (96 kB).  There is no real performance hit from the added second buffer layer, since data is bulk copied in memory from the input buffer to the decoder buffer -- which is orders of magnitude faster than the underlying input IO (especially when this all happens inside the CPU cache -- as it mostly would)...

On a historical note, the IO part of nom-tam-fits harks back to way before the `java.nio` classes appeared, and so its hierarchy is clunky since it originally developed its own high-performance buffering scheme to work with. I've spent a lot of effort to sneak in `java.nio` (and its performance) under the hood however. In the end, the combination of `java.nio` buffering, and the element-by-element translation from there by `FitsDecoder` result in a performance that is not very different from a pure bulk buffer reads from a `java.nio` buffer. I'm afraid there isn't much room left for further improvement without completely redesigning the IO hierarchy (which is a no go for the 1.x line or releases), but rest assurred, the IO performance is not far off from the best you could possibly do even then (based on benchmarks in the tests compared to pure `java.nio` bulk read benchmarks on the same system measured separately)

On a final note, the reason you can't simply create views, such as `FloatBuffer` is because FITS binary data can be wildly heterogeneous (think tables with different column types), and not aligned in any way. So you can have 3 `byte`s of characters, followed by a `double`, followed by a `short`, followed by 5 `float`s... So creating views to read that kind of structure is actually a rabbit hole that is guaranteed to hit you in the knee in terms of performance. Also, while the `FloatBuffer` offers bulk read of `float[]` arrays on the API level, it hides the fact that this is likely implemented element-by-element underneath when reading big-endian data (e.g. FITS) into a little-endian native architecture (such as Intel PCs), which requires byte-reversing each 4-byte chunk individually... (and that would be pretty much equivalent to what `FitsDecoder` does only more explicitly)...

**Comment by keastrid** [comment]
I agree the IO portion of nom.tam is fast, and probably can't be improved. The new nio use added is also great.

Heterogenous structure is an issue, however for images I don't believe it will be a major concern - at least not in so far as `FitsDecoder#readArray`. I'm not seeing any handling for heterogenous structure in its callstack, though I could be missing something. Building the array to return to the user seems to be a bottleneck, particularly for large amounts of pixels. 

For float array bulk get, in the case of mismatched endianness, it calls to native to perform the copy and swap, though I haven't found the source for that yet. This seems to be significantly faster than handling it element-by-element in java when building the array.

A naive implementation (pixel data seems to be complete and unchanged from a quick test) shows significant improvements, so I believe this warrants investigation, at least for `readArray`.

`readArray` changes:
![image](https://github.com/nom-tam-fits/nom-tam-fits/assets/70925169/4919884f-279b-4417-87e1-8326dfe2b1cc)

The following profiles were for a stack of 120 images to emphasize the delta, but I see similar results even for a single image.
Original:
![image](https://github.com/nom-tam-fits/nom-tam-fits/assets/70925169/b32ab945-af94-4aab-978f-7fbbd747cdb8)

Modified:
![image](https://github.com/nom-tam-fits/nom-tam-fits/assets/70925169/3c3d7850-aaf1-4ddc-90d8-964100bac4b9)
Tests performed on Intel i7-6700.

I bring this up as we had a similar issue. ImageJ wants the image to be a 1D array, and in the simple case of `bzero=0` and `bscale=1`, we found using array copies to be significantly faster than looping through all pixels.

**Reply by keastrid** [reply]
This specific change is slower for small amounts of pixels (tested with 51x51 images), but larger images (tested with 2kx2k and 4kx4k) do benefit from this.

**Comment by attipaci** [comment]
Well, I see your point. However, `readArray` is used for reading arrays as small as single element when reading binary tables. Those tables are as commonly used (and overall as large) as typical images. So we have to be careful that while we try to improve image handling performance, we don't screw up binary table performance. 

One option would be to set a minimum array size for triggering reading through a view. However, that will be likely strongly platform dependent, and not going to be future-proof. So, while I'm willing to consider it, I'm a little skeptical if this is really the right way to go... In any case can you check at what element count do the two approaches break even in your benchmarks?

Based on your benchmarks, we are talking a factor of 2 speedup for images. I worry, that we might get a much bigger performance degradation for binary tables with gazillions of small arrays though (which are quite typical). To me being within a factor of 2 of peak performance is 'close enough', and I'm not sure it's worth messing with it at the peril of screwing other applications up...

**Comment by keastrid** [comment]
I want to avoid regressions with small tables, I'm hoping having `InputDecoder` manage the views will be better.

If possible, I'd like to avoid needing a minimum size to trigger the view. 

I didn't see this path get hit for the tables I tried, and it also isn't hit for compressed images, but I'll play around with it and see if I can't get something better.

For us, it is common to see image stacks large enough that opening takes close to or more than a minute, so any improvement is really appreciated.

**Reply by attipaci** [reply]
Try tables with heterogeneous single element columns, e.g. 1 char, 1 short, 1 float, 1 double -- and thousands (millions) of rows...

Also, another approach that might work is to create a new `InputDecoder.readImage()` method that works like `readArray()` but uses views like you suggested. Then `ImageData` can use that variant, while `BinaryTableData` can use the old `readArray` -- each getting what suits them best...

In any case, since this is all under the hood changes -- not affecting the user API, I don't think we need to rush it to make it to 1.18.0. It can go into the next update as well...

**Reply by keastrid** [reply]
A separate method is probably the best.

Yes, there isn't a rush for this.

**Comment by attipaci** [comment, marked as answer]
@keastrid,

I've created PR #414 incorporating the changes you proposed. I do see the expected speedup (~2x for `double[]`) as in your benchmarks, and I think binary tables (and single element) access should be fine also. There may be some low-count element access that may have become slower as a result, but I think in most cases and overall there should be a speedup in general...

Please see if you are happy with the results. If so, I'll merge to 1.18 soon, and release an `rc3` promptly after...
